library/core/src/slice/mod.rs

```diff
@@ -3474,44 +3474,13 @@ impl<T> [T] {
         // Ts = size_of::<U> / gcd(size_of::<T>, size_of::<U>)
         //
         // Luckily since all this is constant-evaluated... performance here matters not!
-        #[inline]
-        fn gcd(a: usize, b: usize) -> usize {
-            use crate::intrinsics;
-            // iterative stein’s algorithm
-            // We should still make this `const fn` (and revert to recursive algorithm if we do)
-            // because relying on llvm to consteval all this is… well, it makes me uncomfortable.
-
-            // SAFETY: `a` and `b` are checked to be non-zero values.
-            let (ctz_a, mut ctz_b) = unsafe {
-                if a == 0 {
-                    return b;
-                }
-                if b == 0 {
-                    return a;
-                }
-                (intrinsics::cttz_nonzero(a), intrinsics::cttz_nonzero(b))
-            };
-            let k = ctz_a.min(ctz_b);
-            let mut a = a >> ctz_a;
-            let mut b = b;
-            loop {
-                // remove all factors of 2 from b
-                b >>= ctz_b;
-                if a > b {
-                    mem::swap(&mut a, &mut b);
-                }
-                b = b - a;
-                // SAFETY: `b` is checked to be non-zero.
-                unsafe {
-                    if b == 0 {
-                        break;
-                    }
-                    ctz_b = intrinsics::cttz_nonzero(b);
-                }
-            }
-            a << k
+        const fn gcd(a: usize, b: usize) -> usize {
+            if b == 0 { a } else { gcd(b, a % b) }
         }
-        let gcd: usize = gcd(mem::size_of::<T>(), mem::size_of::<U>());
+
+        // Explicitly wrap the function call in a const block so it gets
+        // constant-evaluated even in debug mode.
+        let gcd: usize = const { gcd(mem::size_of::<T>(), mem::size_of::<U>()) };
         let ts: usize = mem::size_of::<U>() / gcd;
         let us: usize = mem::size_of::<T>() / gcd;
```

Comment:
Maybe push even a bit more into the const block?
```suggestion
        let (ts, us) = const {
            let gcd: usize = gcd(mem::size_of::<T>(), mem::size_of::<U>());
            (mem::size_of::<U>() / gcd, mem::size_of::<T>() / gcd)
        };
```

Reply:
Doing that results in more instructions being emitted in debug mode for some reason.

Reply:
You can also compute `ts * us` in `const` too.

Reply:
I'm not convinced we should move more operations into the `const`-block if it increases the instruction count though.

Reply:
> Doing that results in more instructions being emitted in debug mode for some reason.

Caveat: small changes in instruction count in a debug build is a subpar proxy for estimating performance of such a build. LLVM can still select instruction sequences that execute in fewer cycles, but require more instructions in a debug mode, so long as it is able to match on IR sequences that trigger those instruction sequences. A prime example of that is e.g. dividing by a constant that produces a strength reduction that involves a multiplication and a shift. Something similar may be happening here.

More importantly, though, this is all LLVM version dependent. It might do better in versions V and V+1, worse in versions V+2 and V+3 and even better in V+4... In this case I'd say, as long as the difference in machine instruction count isn’t like 10 or 20%, then we should strive to ensure that rustc does as much work as possible before handing it off to LLVM’s consideration.

Reply:
If both `T` and `U` are ZSTs then doing the divisions in the const-block causes a compile time divide-by-zero error. `align_to` has an early return if at least one of the types is a ZST, so the divisor will never be zero at runtime, but it looks like the const-block still gets evaluated even though it's technically dead code.

Reply:
Got it, sounds like we have no other choice right now, then. We could look into moving the check for `ts == 0 || us == 0` into the const block too and bail in some other reasonable way, but it probably something that warrants a separate PR anyhow.